Mathlib/Order/SuccPred/Archimedean.lean

```diff
@@ -0,0 +1,282 @@
+
+/-- Induction principle on a type with a `SuccOrder` for all elements above a given element `m`. -/
+@[elab_as_elim]
+theorem Succ.rec {P : α → Prop} {m : α} (h0 : P m) (h1 : ∀ n, m ≤ n → P n → P (succ n)) ⦃n : α⦄
+    (hmn : m ≤ n) : P n := by
+  obtain ⟨n, rfl⟩ := hmn.exists_succ_iterate; clear hmn
+  induction' n with n ih
+  · exact h0
+  · rw [Function.iterate_succ_apply']
+    exact h1 _ (id_le_iterate_of_id_le le_succ n m) ih
+
+theorem Succ.rec_iff {p : α → Prop} (hsucc : ∀ a, p a ↔ p (succ a)) {a b : α} (h : a ≤ b) :
+    p a ↔ p b := by
+  obtain ⟨n, rfl⟩ := h.exists_succ_iterate
+  exact Iterate.rec (fun b => p a ↔ p b) (fun c hc => hc.trans (hsucc _)) Iff.rfl n
+
+lemma le_total_of_ge {r v₁ v₂ : α} (h₁ : r ≤ v₁) (h₂ : r ≤ v₂) : v₁ ≤ v₂ ∨ v₂ ≤ v₁ := by
+  obtain ⟨n, rfl⟩ := h₁.exists_succ_iterate
+  obtain ⟨m, rfl⟩ := h₂.exists_succ_iterate
+  clear h₁ h₂
+  wlog h : n ≤ m
+  · rw [Or.comm]
+    apply this
+    exact Nat.le_of_not_ge h
+  left
+  obtain ⟨k, rfl⟩ := Nat.exists_eq_add_of_le h
+  rw [Nat.add_comm, Function.iterate_add, Function.comp_apply]
+  apply Order.le_succ_iterate
+
+end SuccOrder
+
+section PredOrder
+
+variable [PredOrder α] [IsPredArchimedean α] {a b : α}
+
+instance : IsSuccArchimedean αᵒᵈ :=
+  ⟨fun {a b} h => by convert exists_pred_iterate_of_le h.ofDual⟩
+
+theorem LE.le.exists_pred_iterate (h : a ≤ b) : ∃ n, pred^[n] b = a :=
+  exists_pred_iterate_of_le h
+
+theorem exists_pred_iterate_iff_le : (∃ n, pred^[n] b = a) ↔ a ≤ b :=
+  exists_succ_iterate_iff_le (α := αᵒᵈ)
+
+/-- Induction principle on a type with a `PredOrder` for all elements below a given element `m`. -/
+@[elab_as_elim]
+theorem Pred.rec {P : α → Prop} {m : α} (h0 : P m) (h1 : ∀ n, n ≤ m → P n → P (pred n)) ⦃n : α⦄
+    (hmn : n ≤ m) : P n :=
+  Succ.rec (α := αᵒᵈ) (P := P) h0 h1 hmn
+
+theorem Pred.rec_iff {p : α → Prop} (hsucc : ∀ a, p a ↔ p (pred a)) {a b : α} (h : a ≤ b) :
+    p a ↔ p b :=
+  (Succ.rec_iff (α := αᵒᵈ) hsucc h).symm
+
+lemma le_total_of_le {r v₁ v₂ : α} (h₁ : v₁ ≤ r) (h₂ : v₂ ≤ r) : v₁ ≤ v₂ ∨ v₂ ≤ v₁ :=
+  Or.symm (le_total_of_ge (α := αᵒᵈ) h₁ h₂)
+
+end PredOrder
+
+end Preorder
+
+section PartialOrder
+
+variable [PartialOrder α]
+
+lemma lt_or_le_of_ge [SuccOrder α] [IsSuccArchimedean α] {r v₁ v₂ : α} (h₁ : r ≤ v₁) (h₂ : r ≤ v₂) :
+    v₁ < v₂ ∨ v₂ ≤ v₁ := by
+  rw [Classical.or_iff_not_imp_right]
+  intro nh
+  rcases le_total_of_ge h₁ h₂ with h | h
+  · apply lt_of_le_of_ne h (ne_of_not_le nh).symm
+  · contradiction
+
+/--
+This isn't an instance due to a loop with `LinearOrder`.
+-/
+-- See note [reducible non instances]
+abbrev IsSuccArchimedean.linearOrder [SuccOrder α] [IsSuccArchimedean α]
+     [DecidableEq α] [@DecidableRel α (· ≤ ·)] [@DecidableRel α (· < ·)] [IsDirected α (· ≥ ·)] :
+     LinearOrder α where
+  le_total a b :=
+    have ⟨c, ha, hb⟩ := directed_of (· ≥ ·) a b
+    le_total_of_ge ha hb
+  decidableEq := inferInstance
+  decidableLE := inferInstance
+  decidableLT := inferInstance
```

Comment:
What happens if
```suggestion
```

Reply:
`fields missing: 'decidableLE'`. You can only give it, but then the derived `decidableEq` and `decidableLT` aren't defeq to the given ones.